src/java.base/share/classes/java/lang/StringConcatHelper.java

```diff
@@ -368,16 +368,24 @@ static String simpleConcat(Object first, Object second) {
             // newly created string required, see JLS 15.18.1
             return new String(s1);
         }
-        // start "mixing" in length and coder or arguments, order is not
-        // important
-        long indexCoder = mix(initialCoder(), s1);
-        indexCoder = mix(indexCoder, s2);
-        byte[] buf = newArray(indexCoder);
-        // prepend each argument in reverse order, since we prepending
-        // from the end of the byte array
-        indexCoder = prepend(indexCoder, buf, s2);
-        indexCoder = prepend(indexCoder, buf, s1);
-        return newString(buf, indexCoder);
+        return doConcat(s1, s2);
+    }
+
+    /**
+     * Perform a simple concatenation between two non-empty strings.
+     *
+     * @param s1         first argument
+     * @param s2         second argument
+     * @return String    resulting string
+     */
+    @ForceInline
+    static String doConcat(String s1, String s2) {
+        byte coder = (byte) (s1.coder() | s2.coder());
+        int newLength = (s1.length() + s2.length()) << coder;
+        byte[] buf = newArray(newLength);
+        s1.getBytes(buf, 0, coder);
+        s2.getBytes(buf, s1.length(), coder);
```

Comment:
Does s1 length need to be shifted by coder for dstBegin?

Reply:
Nope - I see the index is char based

https://github.com/openjdk/jdk/blob/69901157e4dae9018abd727956f60fd11b8fa252/src/java.base/share/classes/java/lang/String.java#L4821C15-L4821C23

Reply:
No; in StringUTF16 we abstract away array indices with char index, like in putChar.